### PR TITLE
(tl_mlx5_7) TL/MLX5: ib utilities

### PR DESCRIPTION
**Pull request by vspetrov**
## What
IB utilities of TL/MLX:
* ib context creation
* rc/dc qp creation/connection interfaces

### CODE REVIEW

File: src/components/tl/mlx5/tl_mlx5_ib.c [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+    qp_attr_to_rtr.qp_state          = IBV_QPS_RTR;
+    qp_attr_to_rtr.path_mtu          = IBV_MTU_4096;
+    qp_attr_to_rtr.min_rnr_timer     = 20;
+    qp_attr_to_rtr.ah_attr.port_num  = port_num;
+    qp_attr_to_rtr.ah_attr.is_global = 0;
+
+    qp_attr_to_rts.qp_state      = IBV_QPS_RTS;
+    qp_attr_to_rts.timeout       = 10;
+    qp_attr_to_rts.retry_cnt     = 7;
+    qp_attr_to_rts.rnr_retry     = 7;
```
Comment (Sergei-Lebedev) [inline src/components/tl/mlx5/tl_mlx5_ib.c:286]: i would suggest to make define for numbers e.g.
```
#define UCC_QP_RNR_RETRY 7
#define UCC_QP_TIMEOUT 10
#define UCC_QP_ACCESS_FLAGS (IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_READ | IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_REMOTE_ATOMIC)
```
they are used in multiple places

File: src/components/tl/mlx5/tl_mlx5_ib.c [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+    }
+    // Turn DCI ibv_qp to ibv_qpex and ibv_mqpex
+    dci->dc_qpex = ibv_qp_to_qp_ex(dci->dci_qp);
+    if (!dci->dc_qpex) {
+        tl_error(lib, "Failed turn ibv_qp to ibv_qp_ex, error: %d", errno);
+        goto fail;
+    }
+    dci->dc_mqpex = mlx5dv_qp_ex_from_ibv_qp_ex(dci->dc_qpex);
+    if (!dci->dc_mqpex) {
+        tl_error(lib, "Failed turn ibv_qp_ex to mlx5dv_qp_ex, error: %d",
```
Comment (Sergei-Lebedev) [inline src/components/tl/mlx5/tl_mlx5_ib.c:303]: we usually use small latter in error messages "failed to run ..."

**Review by wfaderhold21 — APPROVED** [review]

**Review by Sergei-Lebedev — APPROVED** [review]

File: src/components/tl/mlx5/tl_mlx5_ib.c [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+
+    memset(&qp_attr, 0, sizeof(qp_attr));
+    qp_attr.qp_state        = IBV_QPS_INIT;
+    qp_attr.pkey_index      = UCC_QP_PKEY_INDEX;
+    qp_attr.port_num        = port;
+    qp_attr.qp_access_flags = UCC_QP_ACCESS_FLAGS;
+    if (ibv_modify_qp(qp, &qp_attr,
+                      IBV_QP_STATE | IBV_QP_PKEY_INDEX | IBV_QP_PORT |
+                          IBV_QP_ACCESS_FLAGS) != 0) {
+        tl_error(lib, "QP RESET->INIT failed");
```
Comment (bureddy) [inline src/components/tl/mlx5/tl_mlx5_ib.c:137]: Add %m format specifier in error logging for all ibv_xx and mlx5dv_xx APIs errors

File: src/components/tl/mlx5/tl_mlx5_ib.c [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+ * See file LICENSE for terms.
+ */
+
+#include "tl_mlx5_ib.h"
+
+#define DC_KEY 1
+#define UCC_QP_RNR_RETRY 7
+#define UCC_QP_RETRY_CNT 7
+#define UCC_QP_RNR_TIMER 20
+#define UCC_QP_TIMEOUT 10
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/mlx5/tl_mlx5_ib.c:13]: added params to config

File: src/components/tl/mlx5/tl_mlx5_ib.c [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+                                   uint8_t port_num, struct ibv_ah **ah_ptr,
+                                   ucc_base_lib_t *lib)
+{
+    struct ibv_ah_attr ah_attr;
+
+    memset(&ah_attr, 0, sizeof(struct ibv_ah_attr));
+
+    ah_attr.dlid          = lid;
+    ah_attr.port_num      = port_num;
+    ah_attr.is_global     = 0;
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/mlx5/tl_mlx5_ib.c:399]: When will we use it? For ROcE? Maybe we should add this as another PR/feature for tl/mlx5

File: src/components/tl/mlx5/tl_mlx5_ib.c [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+    umr_init_attr_ex.cap.max_send_wr  = 1;
+    umr_init_attr_ex.cap.max_recv_wr  = 1;
+    umr_init_attr_ex.cap.max_send_sge = 1;
+    umr_init_attr_ex.cap.max_recv_sge = 1;
+    // `max_inline_data` determines the WQE size that the QP will support.
+    // The 'max_inline_data' should be modified only when the number of
+    // arrays to interleave is greater than 3.
+    //TODO query the devices what is max supported
+    umr_init_attr_ex.cap.max_inline_data =
+        828; // the max number possible, Sergey Gorenko's email
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/mlx5/tl_mlx5_ib.c:442]: we want to grab max supported. i've changed the hardcoded value to "autodetection" - loop qp creation starting from max value of 828 and decreasing it until success.

**Review by bureddy — APPROVED** [review]

**Comment by manjugv** [comment]
This needs input from UCF legal. Blocked until then.

**Comment by manjugv** [comment]
@vspetrov Please add the copyright blurb you want. This was cleared by UCF.

**Pull request merged**